arch/lkl/kernel/traps.c

```diff
@@ -41,7 +41,10 @@ void lkl_do_trap(int trapnr, int signr, char *str, struct ucontext *uctx,
 
 	LKL_TRACE("injecting signal into lkl\n");
 	force_sig_info((struct kernel_siginfo *)info);
-	lkl_process_trap(signr, uctx);
 
 	lkl_cpu_put();
+	
+	/* invoke the signal handler for this trap once we have given up the CPU */
+	/* otherwise we may take the CPU lock again and fail. */
+	lkl_process_trap(signr, uctx);  
```

Comment:
This function calls `get_signal`, which doesn't look safe to call without the CPU lock held.

Reply:
Sorted so that all paths hold the lock and it is only released to service the call back to the user code, then taken again.

Reply:
It is always called with the lock held. I now check that in DEBUG builds and lkl_bug if it is not held in places that need it.